File: src/c/magma/app/locate/BackwardsLocator.c
import magma.api.stream.Stream;import magma.api.stream.Streams;import magma.app.rule.locate.Locator;import java.util.ArrayList;import java.util.List;struct BackwardsLocator{
	String infix;
	public BackwardsLocator(any* _ref_, String infix){
		this.infix =infix;
	}
	String unwrap(any* _ref_){
		return this.infix;
	}
	int length(any* _ref_){
		return this.infix.length();
	}
	Stream<Integer> locate(any* _ref_, String input){
		return Streams.from(searchForIndices(input));
	}
	List<Integer> searchForIndices(any* _ref_, String input){
		List<Integer> indices=new ArrayList<>();
		int index=input.lastIndexOf(this.infix);
		while(index>=0){
			indices.add(index);
			index=input.lastIndexOf(this.infix, index - 1);
		}
		return indices;
	}
	Locator N/A(any* _ref_){
		return N/A.new();
	}
}

File: src/java/magma/app/compile/rule/locate/BackwardsLocator.java
package magma.app.compile.rule.locate;

import magma.api.stream.Stream;
import magma.api.stream.Streams;

import java.util.ArrayList;
import java.util.List;

public class BackwardsLocator implements Locator {
    private final String infix;

    public BackwardsLocator(String infix) {
        this.infix = infix;
    }

    @Override
    public String unwrap() {
        return this.infix;
    }

    @Override
    public int length() {
        return this.infix.length();
    }

    @Override
    public Stream<Integer> locate(String input) {
        return Streams.fromNativeList(searchForIndices(input));
    }

    private List<Integer> searchForIndices(String input) {
        List<Integer> indices = new ArrayList<>();
        int index = input.lastIndexOf(this.infix);
        while (index >= 0) {
            indices.add(index);
            index = input.lastIndexOf(this.infix, index - 1);
        }
        return indices;
    }
}
